### PR TITLE
librbd/cache/pwl: fix buf_persist and add writeback_lat perf counters

### PR DESCRIPTION
**Pull request by CongMinYin**
initialize buf_persist_time, then change name buf_persist_time to
buf_persist_start_time, change flush to internal_flush. add write
back statistics. update some print formats for perf.

before patch:
```
{
    "241be4fc466d": {
        ...
        "op_dis_to_buf_t": {
            "avgcount": 524288,
            "sum": 3370598067.251078760,
            "avgtime": 6428.905615331
        ...
        "op_buf_to_app_t": {
            "avgcount": 524288,
            "sum": 2394009683.918042285,
            "avgtime": 4566.211097560
       ....
}
```
apply patch:
```
""librbd-pwl-10427b9253dc-test-image1"": {
        "op_dis_to_buf_t": {
            "avgcount": 136457,
            "sum": 17.843551539,
            "avgtime": 0.000130763
        },
        "op_buf_to_app_t": {
            "avgcount": 136457,
            "sum": 6.715929409,
            "avgtime": 0.000049216
        },
        "writeback_lat": {
            "avgcount": 136153,
            "sum": 4171.093589702,
            "avgtime": 0.030635341
        },
}
```

Signed-off-by: Yin Congmin <congmin.yin@intel.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by CongMinYin** [comment]
@idryomov As the next step of rwl may be performance tuning, I made some change to perf. Please review.

File: src/librbd/cache/pwl/AbstractWriteLog.cc [hunk 1]
```diff
@@ -258,7 +258,8 @@ void AbstractWriteLog<I>::perf_start(std::string name) {
   plb.add_time_avg(l_librbd_pwl_cmp_latency, "cmp_lat", "Compare and Write latecy");
   plb.add_u64_counter(l_librbd_pwl_cmp_fails, "cmp_fails", "Compare and Write compare fails");
 
-  plb.add_u64_counter(l_librbd_pwl_flush, "flush", "Flush (flush RWL)");
+  plb.add_u64_counter(l_librbd_pwl_internal_flush, "internal_flush", "Flush (write back)");
```
Comment (idryomov) [inline src/librbd/cache/pwl/AbstractWriteLog.cc:261]:
```suggestion
  plb.add_u64_counter(l_librbd_pwl_internal_flush, "internal_flush", "Flush RWL (write back to OSD)");
```

File: src/librbd/cache/pwl/rwl/WriteLog.cc [hunk 2]
```diff
@@ -825,6 +825,16 @@ template <typename I>
 template <typename V>
 void WriteLog<I>::flush_pmem_buffer(V& ops)
 {
+  utime_t now = ceph_clock_now();
+  for (auto &operation : ops) {
+    if (operation->reserved_allocated()) {
+      operation->buf_persist_start_time = now;
```
Comment (idryomov) [inline src/librbd/cache/pwl/rwl/WriteLog.cc:831]: Ah, so `buf_persist_time` wasn't ever set before?

File: src/librbd/cache/pwl/LogOperation.cc [hunk 3]
```diff
@@ -226,12 +226,15 @@ std::ostream &operator<<(std::ostream &os,
 
 void WriteLogOperation::complete(int result) {
   GenericWriteLogOperation::complete(result);
-  m_perfcounter->tinc(l_librbd_pwl_log_op_dis_to_buf_t, buf_persist_time - dispatch_time);
-  utime_t buf_lat = buf_persist_comp_time - buf_persist_time;
-  m_perfcounter->tinc(l_librbd_pwl_log_op_buf_to_bufc_t, buf_lat);
-  m_perfcounter->hinc(l_librbd_pwl_log_op_buf_to_bufc_t_hist, buf_lat.to_nsec(),
+  m_perfcounter->tinc(l_librbd_pwl_log_op_dis_to_buf_t,
+                      buf_persist_start_time - dispatch_time);
+  utime_t buf_persist_time = buf_persist_comp_time - buf_persist_start_time;
```
Comment (idryomov) [inline src/librbd/cache/pwl/LogOperation.cc:231]:
```suggestion
  utime_t buf_persist_lat = buf_persist_comp_time - buf_persist_start_time;
```

**Review by idryomov — CHANGES_REQUESTED** [review]

File: src/librbd/cache/pwl/rwl/WriteLog.cc (hunk 2 above)
Comment (CongMinYin) [reply src/librbd/cache/pwl/rwl/WriteLog.cc:831]: > Ah, so `buf_persist_time` wasn't ever set before?

yes, this fix is the main purpose of this pr.

**Comment by CongMinYin** [comment]
@idryomov done, please review again.

**Comment by CongMinYin** [comment]
jenkins test make check

**Review by idryomov — APPROVED** [review]
Looks good!  Please amend the title of the commit to match the title of the PR.

**Comment by CongMinYin** [comment]
> Looks good! Please amend the title of the commit to match the title of the PR.

Thanks, done.

**Comment by idryomov** [comment]
Since the buf_persist thing is an actual bug, I filed a tracker ticker for backporting purposes.  Could you please add `Fixes: https://tracker.ceph.com/issues/52090` as well?

**Pull request merged**